util/dvsim/NcLauncher.py

```diff
@@ -0,0 +1,225 @@
+# Copyright lowRISC contributors (OpenTitan project).
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import datetime
+import logging as log
+import os
+import shutil
+import subprocess
+from Launcher import ErrorMessage
+from Launcher import Launcher
+from Launcher import LauncherError
+from utils import rm_path
+
+
+class NcLauncher(Launcher):
+    """Implementation of Launcher to launch jobs using altair nc."""
+
+    def __init__(self, deploy):
+        """Initialize common class members."""
+
+        super().__init__(deploy)
+
+        # Popen object when launching the job.
+        self.process = None
+
+    def create_run_sh(self, full_path, cmd):
+        source_file = os.path.join(os.path.dirname(__file__), 'nc_post_cmd.sh')
+        destination_file = os.path.join(full_path, 'nc_post_cmd.sh')
+        shutil.copy2(source_file, destination_file)
+        run_file = os.path.join(full_path, 'run.sh')
+        rm_path(run_file)
+        lines = ['#!/bin/sh',
+                 'function realpath {',
+                 '  python -c "import os; print (os.path.realpath(\'$1\'))"',
+                 '}',
+                 'MY_FILEPATH=$(realpath "${BASH_SOURCE[0]}")',
+                 'MY_DIR=$( dirname "${MY_FILEPATH}" )',
+                 'cd $MY_DIR',
+                 'echo Launch start : `date`',
+                 'SECONDS=0',
+                 cmd,
+                 'echo Launch end : `date`',
+                 'echo CPU time : $SECONDS sec']
+        with open(run_file, 'w') as f:
+            f.write('\n'.join(lines))
+        os.chmod(run_file, 0o755)
+
+    def get_submit_cmd(self, exetool, log_file, job_name, cmd, odir,
+                       interactive_flags):
+        postcmd = (
+            f'{odir}/nc_post_cmd.sh >post.log; '
+            f'cat post.log >>{log_file}'
+        )
+        cmd_array = ['nc', 'run',
+                     '-e', 'SNAPSHOT',
+                     '-nodb', '-forcelog',
+                     '-l', log_file,
+                     '-postcmd', postcmd,
+                     '-set', job_name]
+        cmd_array.extend(interactive_flags)
+        self.create_run_sh(odir, cmd)
+        # TODO: This needs to be moved to a HJson.
+        if (exetool == 'xcelium'):
+            cmd_array.extend(['-r', 'License:Xcelium_Single_Core/1'])
+        if (exetool == 'vcs'):
+            cmd_array.extend(['-r', 'License:VCSRuntime_Net/1'])
+        cmd_array.extend(['--', f'{odir}/run.sh'])
+        return cmd_array
```

Comment:
Thanks for tidying this up! I think things have now got a bit simpler and we can collapse it even further. I *think* this is equivalent and might be easier to follow?
```python3
    def get_submit_cmd(self, exetool, log_file, job_name, cmd, odir,
                       interactive_flags):
        postcmd = (
            f'{odir}/nc_post_cmd.sh >post.log; '
            f'cat post.log >>{log_file}'
        )

        # TODO: These tool-specific names need moving into an hjson config
        # file.
        if (exetool == 'xcelium'):
            license_args = ['-r', 'License:Xcelium_Single_Core/1']
        elif (exetool == 'vcs'):
            license_args = ['-r', 'License:VCSRuntime_Net/1']
        else:
            license_args = []

        self.create_run_sh(odir, cmd)

        return (['nc', 'run',
                 '-e', 'SNAPSHOT',
                 '-nodb', '-forcelog',
                 '-l', log_file,
                 '-postcmd', postcmd,
                 '-set', job_name] +
                license_args +
                interactive_flags +
                ['--', f'{odir}/run.sh'])
```

Reply:
done